src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,268 @@
+
+    /*
+     * How this works:
+     *
+     * Take a rule like this:
+     *
+     * RULE:[2:$1foo$2](Admin\.(.*)s/Admin\.(.*)/$1/
+     *
+     * And say the GSSAPI principal name is:
+     * Admin.site-user/example.com@REALM.com.
+     *
+     * The name breaks down to:
+     * service-name: Admin.site-user
+     * host-name: example.com
+     * realm: REALM.com
+     *
+     * First, the "2" at the beginning ensures there are two parameters in the
+     * principal name (ignoring realm, we have the service-name and host-name).
+     *
+     * If we were missing the host-name, then we would only have one parameter
+     * and this rule would be ignored.
+     *
+     * Now, the "$1foo$2".  This is the 'format'.  This formats the GSSAPI name,
+     * so it can be fed into the next match regex.
+     *
+     * $0 - Realm
+     * $1 - service-name
+     * $2 - host-name
+     *
+     * So this 'format' takes the name and turns it into
+     * "Admin.site-userfooexample.com"
+     *
+     * Next, we test if this formatted name matches the "match" regex which in
+     * this case is "(Admin\.(.*))".  Meaning anything that begins with "Admin".
+     *
+     * It does, so then we move on to the replacement regex:
+     * "s/Admin\.(.*)/$1/".
+     *
+     * The "Admin\.(.*)" is the from pattern.
+     * The "$1" is the to pattern.
+     *
+     * This rule effectively removes the "Admin." at the beginning of the
+     * formatted name:
+     * "site-userfooexample.com" is the result.
+     */
+
+    ss::sstring result;
+
+    if (_is_default) {
+        if (_default_realm == params[0]) {
+            result = ss::sstring(params[1]);
+        }
+    } else if (
+      params.size() - 1 == static_cast<size_t>(_number_of_components)) {
+        auto base = replace_parameters(_format, params);
+        if (!base) {
+            vlog(
+              seclog.error,
+              "GSSAPI: Unable to perform parameter name replacement");
+            return {};
+        }
+        const re2::StringPiece base_piece(base->data(), base->size());
+        if (_match.empty() || RE2::FullMatch(base_piece, match_regex)) {
+            if (!_from_pattern) {
+                result = *base;
+            } else {
+                result = replace_substitution(
+                  *base, *_from_pattern, _to_pattern, _repeat);
+            }
+        }
+    }
+
+    const re2::StringPiece result_piece(result.data(), result.size());
+    if (!result.empty() && RE2::PartialMatch(result_piece, non_simple_regex)) {
+        vlog(
+          seclog.error,
+          "Non-simple name {} after auth_to_local rule {}",
+          result,
+          *this);
+        return {};
+    }
+
+    if (_to_lower && !result.empty()) {
+        boost::algorithm::to_lower(result, std::locale::classic());
+    }
+
+    if (_to_upper && !result.empty()) {
+        boost::algorithm::to_upper(result, std::locale::classic());
```

Comment:
is it true that to_lower and to_upper are mutually exclusive values? if so, maybe use a single bool_class?

Reply:
Yeah, I admit I copied this style from another piece of code.  Using an enum is definitely 'more correct'.

Reply:
if they aren't mutually exclusive, then an enum makes sense. if they are mutually exclusive then it seems like a single bool is sufficient.

Reply:
It's tri-state; upper, lower, ore leave alone.

Reply:
ahh, optional<bool> or enum then?

Reply:
`enum case_change_operation { noop, make_lower, make_upper };`